processing/src/main/java/io/druid/segment/ReferenceCountingSegment.java

```diff
@@ -44,141 +63,74 @@ public ReferenceCountingSegment(Segment baseSegment)
-    return numReferences;
+    return Math.max(referents.getRegisteredParties() - 1, 0);
   }
 
   public boolean isClosed()
   {
-    return isClosed;
+    return referents.isTerminated();
   }
 
   @Override
   public String getIdentifier()
   {
-    synchronized (lock) {
-      if (isClosed) {
-        return null;
-      }
-
-      return baseSegment.getIdentifier();
-    }
+    return !isClosed() ? baseSegment.getIdentifier() : null;
   }
 
   @Override
   public Interval getDataInterval()
   {
-    synchronized (lock) {
-      if (isClosed) {
-        return null;
-      }
-
-      return baseSegment.getDataInterval();
-    }
+    return !isClosed() ? baseSegment.getDataInterval() : null;
   }
 
   @Override
   public QueryableIndex asQueryableIndex()
   {
-    synchronized (lock) {
-      if (isClosed) {
-        return null;
-      }
-
-      return baseSegment.asQueryableIndex();
-    }
+    return !isClosed() ? baseSegment.asQueryableIndex() : null;
   }
 
   @Override
   public StorageAdapter asStorageAdapter()
   {
-    synchronized (lock) {
-      if (isClosed) {
-        return null;
-      }
-
-      return baseSegment.asStorageAdapter();
-    }
+    return !isClosed() ? baseSegment.asStorageAdapter() : null;
   }
 
   @Override
-  public void close() throws IOException
+  public void close()
   {
-    synchronized (lock) {
-      if (isClosed) {
-        log.info("Failed to close, %s is closed already", baseSegment.getIdentifier());
-        return;
-      }
-
-      if (numReferences > 0) {
-        log.info("%d references to %s still exist. Decrementing.", numReferences, baseSegment.getIdentifier());
-
-        decrement();
-      } else {
-        log.info("Closing %s", baseSegment.getIdentifier());
-        innerClose();
-      }
+    if (closed.compareAndSet(false, true)) {
+      referents.arriveAndDeregister();
+    } else {
+      log.warn("close() is called more than once on ReferenceCountingSegment");
     }
   }
 
-  public Closeable increment()
+  public boolean increment()
   {
-    synchronized (lock) {
-      if (isClosed) {
-        return null;
-      }
-
-      numReferences++;
-      final AtomicBoolean decrementOnce = new AtomicBoolean(false);
-      return new Closeable()
-      {
-        @Override
-        public void close() throws IOException
-        {
-          if (decrementOnce.compareAndSet(false, true)) {
-            decrement();
-          }
-        }
-      };
-    }
+    // Negative return from referents.register() means the Phaser is terminated.
+    return referents.register() >= 0;
   }
 
-  private void decrement()
+  public Closeable decrementOnce()
   {
-    synchronized (lock) {
-      if (isClosed) {
-        return;
-      }
-
-      if (--numReferences < 0) {
-        try {
-          innerClose();
-        }
-        catch (Exception e) {
-          log.error("Unable to close queryable index %s", getIdentifier());
-        }
+    AtomicBoolean decremented = new AtomicBoolean(false);
```

Comment:
The more I call decrementOnce, the more times I can close, due to the local variable.